<a href="https://colab.research.google.com/github/mohamadfaisalbashir/grokking-Deep-Learning/blob/main/03_introduction_to_neural_prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Chapter 3: Introduction to Neural Prediction - Forward Propagation**

This notebook summarizes Chapter 3 of *Grokking Deep Learning*:

1. Step 1: Predict
2. A Simple Neural Network Making a Prediction
3. What Is a Neural Network?
4. What Does This Neural Network Do?
5. Making a Prediction with Multiple Inputs
6. Multiple Inputs: What Does This Neural Network Do?
7. Multiple Inputs: Complete Runnable Code
8. Making a Prediction with Multiple Outputs
9. Predicting with Multiple Inputs and Outputs
10. Multiple Inputs and Outputs: How Does It Work?
11. Predicting on Predictions
12. A Quick Primer on NumPy
13. Summary

This is the first chapter in which a neural network is built. It focuses on the first step of the *predict, compare, learn* paradigm introduced earlier, namely **prediction**, and shows how a network turns input data into output predictions. The chapter opens with a remark by Warren Ellis, who jokes that he tries to avoid the business of prediction because it is a quick way to look like an idiot.

---
## **1. Step 1: Predict**

**This chapter is about prediction.**

In the previous chapter, the paradigm of *predict, compare, learn* was introduced. This chapter examines the first step, *predict*, in depth. A prediction involves three parts: the **data**, the **machine**, and the **prediction** itself.

> Data, machine, and prediction for the Red Sox example
>
> <img src="pictures/chapter 3/picture 3.1.png" alt="Figure 3.1" width="600"/>

The first neural network predicts **one datapoint at a time**.

> One datapoint (# toes) enters the machine and a prediction comes out
>
> <img src="pictures/chapter 3/picture 3.2.png" alt="Figure 3.2" width="600"/>

Later, the number of datapoints processed at a time will turn out to have a significant impact on what a network looks like. How many datapoints to propagate at once depends on whether the network can be accurate with the data it receives. For example, to predict whether a photo contains a cat, the network needs all the pixels of the image at once, because a single pixel is not enough to classify it.

This gives a general rule of thumb: **always present enough information to the network**, where "enough information" is defined loosely as how much a human would need to make the same prediction.

A network can only be created once the **shape** of the input and output datasets is understood. For now, shape means the number of columns, or the number of datapoints processed at once. The running example is a single prediction of the likelihood that a baseball team will win.

> toes goes in, win probability comes out
>
> <img src="pictures/chapter 3/picture 3.3.png" alt="Figure 3.3" width="600"/>

Since there is one input datapoint and one output datapoint, the network needs only a **single knob** that maps the input to the output. Abstractly, these knobs are called **weights**.

> An empty network with one weight
>
> <img src="pictures/chapter 3/picture 3.4.png" alt="Figure 3.4" width="600"/>

With one weight, the network takes in one datapoint at a time (the average number of toes per player on the team) and outputs one prediction (whether it thinks the team will win).

---
## **2. A Simple Neural Network Making a Prediction**

**Let's start with the simplest neural network possible.**

The prediction proceeds in four stages, each paired with the corresponding line of code:

1. **An empty network.** The weight and the network function are defined.
2. **Inserting one input datapoint.** The first value of `number_of_toes` is taken as the input.
3. **Multiplying the input by the weight.** The line `prediction = input * weight` computes 8.5 * 0.1 = 0.85.
4. **Depositing the prediction.** The result 0.85 is stored in `pred`.

> The four stages of a simple neural network prediction
>
> <img src="pictures/chapter 3/picture 3.5.png" alt="Figure 3.5" width="600"/>


---
## **3. What Is a Neural Network?**

**Here is your first neural network.** The code below defines the network and then uses it to predict:

In [1]:
weight = 0.1
def neural_network(input, weight):
    prediction = input * weight
    return prediction

number_of_toes = [8.5, 9.5, 10, 9]
input = number_of_toes[0]
pred = neural_network(input,weight)
print(pred)

0.8500000000000001


The last line prints the prediction (`pred`), which should be approximately **0.85**. (Floating-point arithmetic may display it as `0.8500000000000001`.)

For now, a **neural network** is one or more weights that are multiplied by the input data to make a prediction. The author answers four common questions:

- **What is input data?** A number recorded in the real world, usually something easily knowable, such as today's temperature, a baseball player's batting average, or yesterday's stock price.
- **What is a prediction?** What the network tells you given the input data. Examples: "given the temperature, it is 0% likely that people will wear sweatsuits today", "given a batting average, he is 30% likely to hit a home run", or "given yesterday's stock price, today's stock price will be 101.52."
- **Is the prediction always right?** No. A network sometimes makes mistakes, but it can learn from them. If it predicts too high, it adjusts its weight to predict lower next time, and vice versa.
- **How does the network learn?** By trial and error. It first makes a prediction, then checks whether the prediction was too high or too low, and finally changes the weight (up or down) to be more accurate the next time it sees the same input.

---
## **4. What Does This Neural Network Do?**

**It multiplies the input by a weight. It "scales" the input by a certain amount.**

### **4.1 Scaling the Input**

In its simplest form, a neural network uses the power of multiplication. It takes an input datapoint (here 8.5) and multiplies it by the weight.

- If the weight is 2, the network doubles the input.
- If the weight is 0.01, the network divides the input by 100.

Some weight values therefore make the input bigger, and others make it smaller.

> An empty network and inserting one input datapoint
>
> <img src="pictures/chapter 3/picture 3.6.png" alt="Figure 3.6" width="600"/>

### **4.2 The Neural Network Interface**

The interface of a neural network is simple. It accepts an **input** variable as *information* and a **weight** variable as *knowledge*, and it outputs a **prediction**. Every neural network works this way: it uses the knowledge stored in the weights to interpret the information in the input data. Later networks accept larger and more complicated inputs and weights, but the same premise always holds.

In this example, the information is the average number of toes on a baseball team before a game. The network has no access to any information except the current instance. If `number_of_toes[1]` were fed in after the first prediction, the network would not remember the previous prediction. **A neural network knows only what it is fed as input and forgets everything else.** Later, a "short-term memory" is obtained by feeding in multiple inputs at once.

### **4.3 Weight as Sensitivity (the Volume Knob)**

Another way to think about the weight is as a measure of **sensitivity** between the input and the prediction.

- If the weight is very high, even the tiniest input can create a very large prediction.
- If the weight is very small, even large inputs produce small predictions.

This sensitivity is akin to volume. Turning up the weight amplifies the prediction relative to the input, so **the weight is a volume knob**.

> Multiplying input by weight and depositing the prediction
>
> <img src="pictures/chapter 3/picture 3.7.png" alt="Figure 3.7" width="600"/>

In this case, the network applies a volume knob to the `number_of_toes` variable. In theory, that knob could tell the likelihood that the team will win, based on the average number of toes per player. This may or may not work in practice (a team with an average of zero toes would probably play terribly, but baseball is far more complex than that), which motivates presenting multiple pieces of information at once in the next section.

### **4.4 Negative Numbers**

Neural networks are not limited to positive numbers. They can predict negative numbers and can take negative numbers as input. For example, to predict the probability that people will wear coats today, a temperature of -10 degrees Celsius combined with a negative weight yields a high probability.

> Temperature to probability with a negative weight
>
> <img src="pictures/chapter 3/picture 3.8.png" alt="Figure 3.8" width="600"/>


---
## **5. Making a Prediction with Multiple Inputs**

**Neural networks can combine intelligence from multiple datapoints.**

The previous network took one datapoint and made one prediction from it. The natural question is whether the average number of toes is really a good predictor by itself. If the network were given more information at one time, it should, in theory, make more accurate predictions. A network can accept **multiple input datapoints at a time**.

The network now takes three inputs per prediction:

- `toes`: the current average number of toes per player
- `wlrec`: the current games won (as a percentage)
- `nfans`: the fan count (in millions)

The dataset records the status at the beginning of each of the first four games of a season, and the input for the first game is built from the first entry of each list.

> An empty network with multiple inputs, and inserting one row of data
>
> <img src="pictures/chapter 3/picture 3.9.png" alt="Figure 3.9" width="600"/>


The prediction is computed in two further stages: a **weighted sum of the inputs**, followed by depositing the result.

> Performing a weighted sum of inputs and depositing the prediction
>
> <img src="pictures/chapter 3/picture 3.10.png" alt="Figure 3.10" width="600"/>


In [2]:
def w_sum(a,b):
    assert(len(a) == len(b))
    output = 0
    for i in range(len(a)):
        output += (a[i] * b[i])
    return output

weights = [0.1, 0.2, 0]

def neural_network(input, weights):
    pred = w_sum(input,weights)
    return pred

toes =  [8.5, 9.5, 9.9, 9.0]
wlrec = [0.65, 0.8, 0.8, 0.9]
nfans = [1.2, 1.3, 0.5, 1.0]

input = [toes[0],wlrec[0],nfans[0]]
pred = neural_network(input,weights)

print(pred)

0.9800000000000001


---
## **6. Multiple Inputs: What Does This Neural Network Do?**

**It multiplies three inputs by three knob weights and sums them. This is a weighted sum.**

The limiting factor of the earlier network was that it was only a volume knob on a single datapoint. To make accurate predictions, a network must **combine multiple inputs at the same time**, and this is exactly what the new network does.

The fundamental mechanism has not changed: every input still passes through its own volume knob, meaning each input is multiplied by its own weight. The new property is that, with multiple inputs, the individual (local) predictions must be **summed**. This is called a **weighted sum of the input**, or simply a **weighted sum**. It is also referred to as a **dot product**.

The same four-stage figures as in Section 5 apply here (an empty network, inserting one row of data, performing the weighted sum, and depositing the prediction), so Figures 3.9 and 3.10 are reused.

> **A relevant reminder:** the interface of the network is simple. It accepts an `input` variable as information and a `weights` variable as knowledge, and it outputs a prediction.

### **6.1 Vectors and Elementwise Operations**

Processing multiple inputs at once justifies a new tool: the **vector**. A vector is nothing other than a **list of numbers**. In the example, `input` and `weights` are both vectors (and three more vectors appear in the code: `toes`, `wlrec`, and `nfans`).

Vectors are useful whenever operations involve groups of numbers. A weighted sum is an operation between two vectors of equal length: each number is multiplied by the number in the same position of the other vector (position 0 with 0, position 1 with 1, and so on), and the products are then summed.

Any mathematical operation between two vectors of equal length that pairs up values by position is called an **elementwise operation**. Hence:

- **Elementwise addition** sums two vectors.
- **Elementwise multiplication** multiplies two vectors.

> **Challenge: vector math.** Manipulating vectors is a cornerstone technique for deep learning. Try writing functions that perform the following operations, and then use two of them to perform a dot product:
>
> - `def elementwise_multiplication(vec_a, vec_b)`
> - `def elementwise_addition(vec_a, vec_b)`
> - `def vector_sum(vec_a)`
> - `def vector_average(vec_a)`

### **6.2 The Intuition Behind the Dot Product**

Understanding why a dot product works is one of the most important parts of understanding how neural networks make predictions. Loosely stated, a dot product gives a **notion of similarity** between two vectors. Consider these examples:

```
a = [ 0, 1, 0, 1]
b = [ 1, 0, 1, 0]
c = [ 0, 1, 1, 0]
d = [.5, 0,.5, 0]
e = [ 0, 1,-1, 0]

w_sum(a,b) = 0
w_sum(b,c) = 1
w_sum(b,d) = 1
w_sum(c,c) = 2
w_sum(d,d) = .5
w_sum(c,e) = 0
```

The highest weighted sum, `w_sum(c,c)`, is between vectors that are exactly identical. By contrast, `a` and `b` have no overlapping weight, so their dot product is zero.

The most interesting case is between `c` and `e`: the negative weight in `e` cancels the positive similarity between them. A dot product of `e` with itself would still yield 2, because a double negative turns positive.

### **6.3 The Dot Product as a Logical AND**

The properties of the dot product can sometimes be equated to a **logical AND**.

```
a = [ 0, 1, 0, 1]
b = [ 1, 0, 1, 0]
```

Do both `a[0]` AND `b[0]` have value? No. Do both `a[1]` AND `b[1]` have value? No again. Since this holds for all four positions, the final score is 0: every position fails the logical AND.

```
b = [ 1, 0, 1, 0]
c = [ 0, 1, 1, 0]
```

`b` and `c` share value in one column (`b[2]` and `c[2]`), so that column alone passes the AND and lifts the score to 1.

```
c = [ 0, 1, 1, 0]
d = [.5, 0,.5, 0]
```

Neural networks can also model **partial ANDing**. `c` and `d` share the same column as `b` and `c`, but `d` has only 0.5 weight there, so the score is only 0.5. This property is exploited when modeling probabilities in neural networks.

```
d = [.5, 0,.5, 0]
e = [-1, 1, 0, 0]
```

**Negative weights** tend to imply a logical **NOT**, since any positive weight paired with a negative weight pulls the score down. If both vectors carry negative weights (as in `w_sum(e,e)`), a double negative occurs and the score increases instead. It can also be viewed as an **OR after the AND**: if any position shows weight, the score is affected. For `w_sum(a,b)`, a positive score results if (`a[0]` AND `b[0]`) OR (`a[1]` AND `b[1]`), and so on.

### **6.4 A Crude Language for Reading Weights**

These intuitions give a crude language for reading weights. Assume `w_sum(input, weights)` is being computed and the "then" of each statement is an abstract "then give a high score":

```
weights = [ 1, 0, 1]   => if input[0] OR input[2]
weights = [ 0, 0, 1]   => if input[2]
weights = [ 1, 0, -1]  => if input[0] OR NOT input[2]
weights = [-1, 0, -1]  => if NOT input[0] OR NOT input[2]
weights = [ 0.5, 0, 1] => if BIG input[0] or input[2]
```

In the last row, `weight[0] = 0.5` means the corresponding `input[0]` must be larger to compensate for the smaller weighting. The language is very approximate, but it is helpful for picturing what happens under the hood, especially as networks become more complex.

Roughly speaking, then, **a network gives a high score to inputs based on how similar they are to its weights**.

### **6.5 Reading the Example Result**

In the example, `nfans` is completely ignored because its weight is 0. The most sensitive predictor is `wlrec`, since its weight (0.2) is the highest. Nevertheless, the dominant force behind the high score is the number of toes (`toes`), not because its weight is the highest, but because the input combined with the weight is by far the highest.

Further points to remember:

- Weights **cannot be shuffled**: each has a specific position it must occupy.
- **Both** the value of the weight and the value of the input determine the impact on the final score.
- A negative weight causes some inputs to reduce the final prediction (and vice versa).

---
## **7. Multiple Inputs: Complete Runnable Code**

The code from the previous sections (written with basic Python lists and numbers) is the version shown in Section 5. A better way exists and will be used from now on: the **NumPy** library ("numerical Python"), which has very efficient code for creating vectors and performing common operations such as dot products.

The NumPy version below does not need a `w_sum` function, because NumPy already provides `dot` (short for "dot product"). Both versions should print approximately **0.98**. Many functions that will be used later have NumPy equivalents.

In [3]:
import numpy as np
weights = np.array([0.1, 0.2, 0])
def neural_network(input, weights):
    pred = input.dot(weights)
    return pred

toes =  np.array([8.5, 9.5, 9.9, 9.0])
wlrec = np.array([0.65, 0.8, 0.8, 0.9])
nfans = np.array([1.2, 1.3, 0.5, 1.0])

input = np.array([toes[0],wlrec[0],nfans[0]])
pred = neural_network(input,weights)

print(pred)

0.9800000000000001


---
## **8. Making a Prediction with Multiple Outputs**

**Neural networks can also make multiple predictions using only a single input.**

Multiple outputs are a simpler augmentation than multiple inputs. Prediction occurs as if there were **three disconnected single-weight networks**. Here, instead of predicting only whether the team will win, the network also predicts whether the players are happy or sad and the percentage of team members who are hurt, using only the current win/loss record.

The key point is that the three predictions are **completely separate**. Unlike a network with multiple inputs and one output, where the prediction is undeniably connected, this network behaves as three independent components that each receive the same input. This makes it simple to implement.

> An empty network with multiple outputs, and inserting one input datapoint
>
> <img src="pictures/chapter 3/picture 3.11.png" alt="Figure 3.11" width="600"/>

> Elementwise multiplication and depositing predictions
>
> <img src="pictures/chapter 3/picture 3.12.png" alt="Figure 3.12" width="600"/>


In [4]:
def ele_mul(number,vector):
    output = [0,0,0]
    assert(len(output) == len(vector))
    for i in range(len(vector)):
        output[i] = number * vector[i]
    return output

weights = [0.3, 0.2, 0.9]

def neural_network(input, weights):
    pred = ele_mul(input,weights)
    return pred

wlrec = [0.65, 0.8, 0.8, 0.9]
input = wlrec[0]
pred = neural_network(input,weights)

print(pred)

[0.195, 0.13, 0.5850000000000001]


---
## **9. Predicting with Multiple Inputs and Outputs**

**Neural networks can predict multiple outputs given multiple inputs.**

The techniques for multiple inputs and for multiple outputs can be combined into a network with both. As before, a weight connects **each input node to each output node**, and prediction proceeds in the usual way.

> An empty network with multiple inputs and outputs, and inserting one row of data
>
> <img src="pictures/chapter 3/picture 3.13.png" alt="Figure 3.13" width="600"/>

> For each output, a weighted sum of inputs, and depositing predictions
>
> <img src="pictures/chapter 3/picture 3.14.png" alt="Figure 3.14" width="600"/>


In [5]:
            #toes %win #fans
weights = [ [0.1, 0.1, -0.3], #hurt?
            [0.1, 0.2, 0.0], #win?
            [0.0, 1.3, 0.1] ] #sad?

def w_sum(a,b):
    assert(len(a) == len(b))
    output = 0
    for i in range(len(a)):
        output += (a[i] * b[i])
    return output

def vect_mat_mul(vect,matrix):
    assert(len(vect) == len(matrix))
    output = [0,0,0]
    for i in range(len(vect)):
        output[i] = w_sum(vect,matrix[i])
    return output

def neural_network(input, weights):
    pred = vect_mat_mul(input,weights)
    return pred

toes =  [8.5, 9.5, 9.9, 9.0]
wlrec = [0.65,0.8, 0.8, 0.9]
nfans = [1.2, 1.3, 0.5, 1.0]

input = [toes[0],wlrec[0],nfans[0]]
pred = neural_network(input,weights)

print(pred)

[0.555, 0.9800000000000001, 0.9650000000000001]


---
## **10. Multiple Inputs and Outputs: How Does It Work?**

**It performs three independent weighted sums of the input to make three predictions.**

This architecture can be viewed from two perspectives:

1. Three weights coming **out of** each input node, or
2. Three weights going **into** each output node.

The second perspective is much more useful for now. The network is then seen as **three independent dot products**: each output node takes its own weighted sum of the input and makes its own prediction. The four-stage figures are the same as in Section 9, so Figures 3.13 and 3.14 apply again.

The code introduces a new function, `vect_mat_mul`. It iterates through each row of the weights (each row is a vector) and makes a prediction with `w_sum`. In effect, it performs three consecutive weighted sums and stores their predictions in a vector called `output`. Although more weights are involved, this network is not much more advanced than the previous ones.

### **Matrices and Vector-Matrix Multiplication**

The `weights` variable is a **list of vectors**, and a list of vectors is called a **matrix**. Commonly used functions operate on matrices. One of them is **vector-matrix multiplication**: take a vector and perform a dot product with every row of a matrix. The series of weighted sums performed above is exactly this operation. NumPy provides special functions to help with it.

*Note:* for readers experienced in linear algebra, the more formal definition stores and processes weights as column vectors rather than row vectors. The book states that this will be rectified shortly; it is why the `.T` (transpose) appears in the NumPy code of the next section.

---
## **11. Predicting on Predictions**

**Neural networks can be stacked!**

The output of one network can also be fed as the input of another network. This results in **two consecutive vector-matrix multiplications**. It may not yet be clear why one would predict this way, but some datasets (such as image classification) contain patterns that are too complex for a single weight matrix. The nature of these patterns is discussed later; for now it suffices to know that stacking is possible.

The middle layer is called the **hidden layer** (`hid`), and the network has three layers: **inputs**, **hiddens**, and **predictions**. Two weight matrices are used: `ih_wgt` (input to hidden) and `hp_wgt` (hidden to prediction). In basic Python, the network is written as follows (as shown in the book's figure):

```python
            # toes % win # fans
ih_wgt = [  [0.1, 0.2, -0.1],  # hid[0]
            [-0.1, 0.1, 0.9],  # hid[1]
            [0.1, 0.4, 0.1] ]  # hid[2]

            # hid[0] hid[1] hid[2]
hp_wgt = [  [0.3, 1.1, -0.3],  # hurt?
            [0.1, 0.2, 0.0],   # win?
            [0.0, 1.3, 0.1] ]  # sad?

weights = [ih_wgt, hp_wgt]

def neural_network(input, weights):
    hid = vect_mat_mul(input, weights[0])
    pred = vect_mat_mul(hid, weights[1])
    return pred
```

> A stacked network: empty network and predicting the hidden layer
>
> <img src="pictures/chapter 3/picture 3.15.png" alt="Figure 3.15" width="600"/>

> Predicting the output layer and depositing the prediction
>
> <img src="pictures/chapter 3/picture 3.16.png" alt="Figure 3.16" width="600"/>

The listing below performs the same operations using NumPy, which makes the code faster and easier to read and write. Note the `.T` after each weight matrix: it transposes the matrix so that each column, rather than each row, holds the weights feeding one node.

In [6]:
import numpy as np

#toes %win #fans
ih_wgt = np.array([
            [0.1, 0.2, -0.1], #hid[0]
            [-0.1,0.1, 0.9], #hid[1]
            [0.1, 0.4, 0.1]]).T #hid[2]


# hid[0] hid[1] hid[2]
hp_wgt = np.array([
            [0.3, 1.1, -0.3], #hurt?
            [0.1, 0.2, 0.0], #win?
            [0.0, 1.3, 0.1] ]).T #sad?

weights = [ih_wgt, hp_wgt]

def neural_network(input, weights):

    hid = input.dot(weights[0])
    pred = hid.dot(weights[1])
    return pred


toes =  np.array([8.5, 9.5, 9.9, 9.0])
wlrec = np.array([0.65,0.8, 0.8, 0.9])
nfans = np.array([1.2, 1.3, 0.5, 1.0])

input = np.array([toes[0],wlrec[0],nfans[0]])

pred = neural_network(input,weights)
print(pred)

[0.2135 0.145  0.5065]


---
## **12. A Quick Primer on NumPy**

**NumPy does a few things for you. Let's reveal the magic.**

This chapter introduced two new mathematical tools, **vectors** and **matrices**, together with several operations on them: dot products, elementwise multiplication and addition, and vector-matrix multiplication. So far these operations were written as Python functions that work on plain `list` objects. Those functions will keep being used for a while so that what happens inside them is fully understood. This primer prepares the reader for the transition to NumPy-only chapters.

### **12.1 Creating Vectors and Matrices**

Vectors and matrices can be created in several ways in NumPy. Creating a vector and creating a matrix follow the same process; a matrix with only one row is effectively a vector. As in mathematics, the shape of a matrix is written as **(rows, columns)**: rows come first, columns second.

In [7]:
import numpy as np

a = np.array([0,1,2,3])
b = np.array([4,5,6,7])
c = np.array([[0,1,2,3],
              [4,5,6,7]])

d = np.zeros((2,4))
e = np.random.rand(2,5)

print(a)
print(b)
print(c)
print(d)
print(e)

[0 1 2 3]
[4 5 6 7]
[[0 1 2 3]
 [4 5 6 7]]
[[0. 0. 0. 0.]
 [0. 0. 0. 0.]]
[[0.21769126 0.47007385 0.79123029 0.42455058 0.97165162]
 [0.72746778 0.63580947 0.02405697 0.15448172 0.96641849]]


### **12.2 NumPy's Automatic Operation Detection**

When two variables are multiplied with `*`, NumPy automatically detects what kinds of variables they are and works out which operation is meant. This is very convenient but can make NumPy code harder to read, so the type of each variable must be tracked.

**General rule for anything elementwise (+, -, *, /):** either the two variables must have the *same* number of columns, or one of them must have only one column.

- `print(a * 0.1)` is **vector-scalar multiplication**: the scalar 0.1 multiplies every value in the vector.
- `print(c * 0.2)` looks the same, but NumPy knows `c` is a matrix, so it performs **scalar-matrix multiplication**, multiplying every element by 0.2. A scalar has only one column, so it can be multiplied (or divided, added, subtracted) with anything.
- `print(a * b)` is **elementwise vector multiplication**. Neither vector has only one column, so NumPy checks that they have the same number of columns; since they do, each element is multiplied by the element in the same position. The same holds for addition, subtraction, and division.
- `print(a * c)` is perhaps the most elusive. `a` is a vector with four columns and `c` is a (2 x 4) matrix. They have the same number of columns, so NumPy multiplies the vector `a` with each row of `c`, as if doing elementwise vector multiplication on every row.
- `print(a * e)` raises an error, because `a` (4 columns) and `e` (a 2 x 5 matrix) do not have the same number of columns.

The first cell below runs the first five of these operations. The last operation, `print(a * e)`, is placed in a cell of its own because it is expected to raise a `ValueError`.

In [8]:
print(a * 0.1)

print(c * 0.2)

print(a * b)

print(a * b * 0.2)

print(a * c)

[0.  0.1 0.2 0.3]
[[0.  0.2 0.4 0.6]
 [0.8 1.  1.2 1.4]]
[ 0  5 12 21]
[0.  1.  2.4 4.2]
[[ 0  1  4  9]
 [ 0  5 12 21]]


In [9]:
print(a * e)

ValueError: operands could not be broadcast together with shapes (4,) (2,5) 

### **12.3 Reading NumPy Code and the Golden Rule of `dot`**

All of these operations look the same if it is unknown which variables are scalars, vectors, or matrices. When "reading NumPy", two things are being done: reading the operations, and keeping track of the **shape** (number of rows and columns) of every variable. This takes practice, but it eventually becomes second nature.

Here is a first example of matrix multiplication, noting the input and output shapes: a vector of length 4 is dotted with a matrix of 4 rows and 3 columns.

In [10]:
a = np.zeros((1,4))
b = np.zeros((4,3))

c = a.dot(b)
print(c.shape)

(1, 3)


**The golden rule when using `dot`:** if the `(rows, cols)` descriptions of the two variables being "dotted" are written next to each other, the neighboring numbers must be the same. Here, (1,4) is dotted with (4,3), which works and outputs (1,3).

Regardless of whether vectors or matrices are being dotted, their shapes must line up: the columns of the left matrix must equal the rows of the right matrix.

```
(a,b).dot(b,c) = (a,c)
```

The following cell shows three valid examples. The `.T` operation flips the rows and columns of a matrix.

In [11]:
a = np.zeros((2,4))
b = np.zeros((4,3))

c = a.dot(b)
print(c.shape)

e = np.zeros((2,1))
f = np.zeros((1,3))

g = e.dot(f)
print(g.shape)

h = np.zeros((5,4)).T
i = np.zeros((5,6))

j = h.dot(i)
print(j.shape)

(2, 3)
(2, 3)
(4, 6)


When the shapes do not line up, NumPy raises an error. In the next cell, (5,4) is dotted with (5,6): the neighboring numbers 4 and 5 differ, so the operation fails.

In [12]:
h = np.zeros((5,4))
i = np.zeros((5,6))
j = h.dot(i)
print(j.shape)

ValueError: shapes (5,4) and (5,6) not aligned: 4 (dim 1) != 5 (dim 0)

---
## **13. Summary**

**To predict, neural networks perform repeated weighted sums of the input.**

Chapter 3 presents an increasingly complex variety of neural networks, all built from a relatively small number of simple rules that are repeated and stacked. The key takeaways are:

- A **neural network**, in its simplest form, is one or more **weights** that are multiplied by the **input** data to produce a **prediction**. The interface is always the same: input (information) and weights (knowledge) go in, and a prediction comes out.
- A single weight acts as a **volume knob** that scales the input. Weights can be positive or negative, and so can inputs and predictions.
- With **multiple inputs**, the network computes a **weighted sum** (dot product) of the inputs. The dot product gives a notion of similarity between the input and the weights, and can be read through crude analogies to logical AND, OR, and NOT.
- With **multiple outputs**, the network behaves as several independent single-weight networks that share the same input.
- With **multiple inputs and outputs**, the network performs several independent weighted sums, which is a **vector-matrix multiplication**. A list of vectors is a **matrix**.
- Networks can be **stacked**: the output of one layer becomes the input to the next, giving two consecutive vector-matrix multiplications through a **hidden layer**.
- **NumPy** performs vectors, matrices, and these operations efficiently. It is essential to keep track of the **shape** of every variable, and the golden rule of `dot` is that the neighboring dimensions must match: `(a,b).dot(b,c) = (a,c)`.

Everything in this chapter is a form of **forward propagation**: a neural network takes input data and makes a prediction. It is called this because **activations** are propagated *forward* through the network. In these examples, activations are all the numbers that are *not* weights and that are unique to every prediction.

The intelligence of a network depends on the values of its weights. In the next chapter, the focus moves to **how to set weights** so that networks make accurate predictions. Just as prediction is built from several simple techniques that are repeated and stacked, weight learning is also a series of simple techniques combined many times across an architecture.